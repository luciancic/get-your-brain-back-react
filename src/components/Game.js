import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import GameBoard from './GameBoard';
import * as actions from '../actions/gameActions';

class Game extends Component {
    componentDidMount() {
        // Do not destructure game because its values will not stay up to date in the interval
        const { settings, startRound, endGame } = this.props;

        startRound();
        this.interval = setInterval(() => {
            if (this.props.game.currentRound === settings.maxRounds + settings.n) {
                return endGame();
            }
            startRound();
        }, settings.duration);
    }
    
    componentWillUnmount() {
        const { game, cancelGame } = this.props;

        clearInterval(this.interval);
        if (!game.ended) { cancelGame() }
    }
    
    render() {
        const { n, maxRounds } = this.props.settings;
        const { currentRound, ended, positions } = this.props.game;
        const { round } = this.props;
        const remainingRounds = maxRounds - currentRound + n + 1;

        return <div className="container">
            <div style={{ display: 'block', marginTop: '15px' }}>
                <span className='left' style={{ paddingLeft: '15px'}}>Remaining rounds: { remainingRounds }</span>
                <span className='right' style={{ paddingRight: '15px'}}>N-back: {n}</span>
            </div>
            <GameBoard active={round.active} currentPosition={positions[0]}/>
            { ended ? <Redirect to="/" /> : null }
        </div>
    }
}

function mapStateToProps({ settings, game, round }) {
    return { settings, game, round }
}

export default connect(mapStateToProps, actions)(Game)
