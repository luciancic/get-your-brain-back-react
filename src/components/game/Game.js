import React, { Component } from 'react';
import { connect } from 'react-redux';
import Court from '../court/Court';
import { startGame, endGame, cancelGame } from '../../actions/gameActions';
import { startRound, endRound, answer, playAudio } from '../../actions/roundActions';

import './Game.css';

class Game extends Component {
    componentDidMount() {
        const { endGame } = this.props;
        const { gameRunning } = this.props.game;
        const { duration, maxRounds, n } = this.props.settings;

        // First round starts immediately.
        if (gameRunning) this.playRound();

        // Set up a round loop.
        this.interval = setInterval(() => {
            // Do not destructure currentRound because it needs to check value by reference.
            if (this.props.game.currentRound === maxRounds + n) {
                return endGame();
            }
            this.playRound();
        }, duration + 300);

        window.addEventListener('keydown', this.gameListener, false);
    }
    
    componentWillUnmount() {
        const { game, cancelGame } = this.props;
        
        clearInterval(this.interval);
        clearTimeout(this.timeout);
        window.removeEventListener('keydown', this.gameListener, false);
        
        // In case component unmounts because of back navigation, we need to cancel manually.
        if (game.gameRunning) { cancelGame() }
    }
    
    gameListener = (e) => {
        switch (e.key) {
            case 'Escape':
                this.props.cancelGame();
                break;
            case 'a':
                this.match('positions')();
                break;
            case 'l':
                this.match('letters')();
                break;
            default:
                break;
        }
    }

    match = (stimulus) => () => {
        const { answer, settings } = this.props;
        const { answers, currentRound, roundActive } = this.props.game;
        const presentArray = this.props.game[stimulus];
        
        if (answers[stimulus] === null && roundActive && currentRound > settings.n) {
            answer({ [stimulus]: presentArray[0] === presentArray[settings.n]})
        }
    }

    playRound = () => {
        const { startRound, endRound } = this.props;
        const { n, duration } = this.props.settings;
        const { positions, letters } = this.props.game;
        startRound(positions[n - 1], letters[n - 1]);
        this.timeout = setTimeout(() => {
            // Must destructure values from game inside the setTimeout to ensure the values are extracted at the correct moment. 
            const { answers, positions, letters } = this.props.game;
            endRound({ 
                positionMissed: answers.positions === null && positions[0] === positions[n], 
                letterMissed: answers.letters === null && letters[0] === letters[n]
            });
        }, duration);
    }

    getFeedbackColor(feedback) {
        switch (feedback) {
            case null: return 'white';
            case 'correct': return 'green';
            case 'mistake': return 'red';
            case 'missed': return 'blue';
            default: break;
        }
    }

    render() {
        const { audioPlayed, game, settings } = this.props;
        const { n, maxRounds } = settings;
        const { gameRunning, roundActive, currentRound } = game;
        const remainingRounds = maxRounds - currentRound + n + 1;
        const positionsButtonColor = this.getFeedbackColor(this.props.feedback.positions);
        const lettersButtonColor = this.getFeedbackColor(this.props.feedback.letters);
        
        const { positions, letters } = this.props.game;
        return <Court 
            audioPlayed={audioPlayed}
            gameRunning={gameRunning}
            letters={letters}
            lettersButtonColor={lettersButtonColor}
            lettersHandler={this.match('letters')}
            maxRounds={maxRounds}
            n={n}
            playAudio={this.props.playAudio}
            positions={positions}
            positionHandler={this.match('positions')}
            positionsButtonColor={positionsButtonColor}
            remainingRounds={remainingRounds}
            roundActive={roundActive}
        />
    }
}

function mapStateToProps({ settings, game, feedback, audioPlayed }) {
    return { settings, game, feedback, audioPlayed }
}

const actions = {
    answer,
    startGame, 
    endGame, 
    cancelGame,
    startRound,
    endRound,
    playAudio
}

export default connect(mapStateToProps, actions)(Game)
