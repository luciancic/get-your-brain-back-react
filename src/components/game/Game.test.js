import React from 'react';
import { shallow } from 'enzyme';
import { Game } from './Game';

it('renders Game', () => {
  const wrapper = <Game settings={{}} game={{}} feedback={{}} answer={ jest.fn() }/>;
  expect( shallow( wrapper )).toMatchSnapshot();
});