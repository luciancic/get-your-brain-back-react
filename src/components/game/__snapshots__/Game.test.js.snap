// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Game 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Game
    answer={[MockFunction]}
    feedback={Object {}}
    game={Object {}}
    settings={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <escapeable
          lettersHandler={[Function]}
          positionHandler={[Function]}
          remainingRounds={NaN}
        />,
        <Redirect
          push={false}
          to="/"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "audioPlayed": undefined,
          "letters": undefined,
          "lettersButtonColor": undefined,
          "lettersHandler": [Function],
          "maxRounds": undefined,
          "n": undefined,
          "playAudio": undefined,
          "positionHandler": [Function],
          "positions": undefined,
          "positionsButtonColor": undefined,
          "remainingRounds": NaN,
          "roundActive": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "push": false,
          "to": "/",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <escapeable
            lettersHandler={[Function]}
            positionHandler={[Function]}
            remainingRounds={NaN}
          />,
          <Redirect
            push={false}
            to="/"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "audioPlayed": undefined,
            "letters": undefined,
            "lettersButtonColor": undefined,
            "lettersHandler": [Function],
            "maxRounds": undefined,
            "n": undefined,
            "playAudio": undefined,
            "positionHandler": [Function],
            "positions": undefined,
            "positionsButtonColor": undefined,
            "remainingRounds": NaN,
            "roundActive": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "push": false,
            "to": "/",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
